ios/viewmodels/view-model-reuse.md

```diff
@@ -0,0 +1,167 @@
+
+
+### Protocol based view model
+
+We can use protocols to standardize the interface for the view, while keeping implementation details abstracted out into it's own entity.
+
+```swift
+
+protocol TitleMessageCellConforming {
+  let title: String { get }
+  let message: String { get }
+  let showsEditButton: Bool { get }
+}
+
+```
+
+Note that the only difference between the protocol and the concrete implementation of this class is the type. With this, the details of each cell in ReviewViewController can be abstracted out.
+
+```swift
+
+// PhoneNumberTitleMessageCellViewModel.swift
+struct PhoneNumberTitleMessageCellViewModel: TitleMessageCellConforming {
+  let title: String
+  let message: String
+  let showsEditButton: Bool
+
+  init(profile: Profile) {
+    let formatter = PhoneNumberFormatter(forCountryCode: currentCountryCode)
+    let number = formatter.formatNumber(profile.phoneNumber) ?? "-"
+    title = "Contact number"
+    message = number
+    showsEditButton = profile.canEdit
+  }
+
+// TopicsTitleMessageCellViewModel.swift
+struct TopicsTitleMessageCellViewModel: TitleMessageCellConforming {
+  let title: String
+  let message: String
+  let showsEditButton: Bool
+
+  init(topics: [Topics], review: Review) {
+    title = "Visit Details"
+    message = topics.filter({ $0.selected }).map({ $0.title }).joinWithSeparator("\n")
+    showsEditButton = review.canEdit
+  }
+}
+
+```
+
+This moves view logic code into the view models, leaving a very nice and clean implementation for our view controller:
+
+```swift
+
+// ReviewViewController.swift
+func viewModel(forIndex index: Int) {
+  switch index {
+    case 0:
+      return PhoneNumberTitleMessageCellViewModel(profile)
+    case 1:
+      return AllergiesTitleMessageCellViewModel(allergies, review: review)
+
+    ...
+  }
+}
+
+```
+
+For the profile, it might still be okay to leave the view model logic in the profile view controller, due to it's simplicity.
+
+```swift
+
+// ProfileViewController.swift
+func viewModel(forIndex index: Int) {
+  switch index {
+    case 0:
+      return TitleMessageCellViewModel(title: "First Name", message: profile.firstName) // Now, TitleMessageCellViewModel is simply a struct that implements TitleMessageCellConforming.
+    case 1:
+      return TitleMessageCellViewModel(title: "Surname", message: profile.lastName)
+    ...
+  }
+}
+
+```
+
+## Conclusions and takeaways
+
+View models are a tool to facilitate view logic abstraction, but also code reuse. A protocol-based approach allow us to take abstraction one level further, so that our view relies on a generic interface, without forcing an implementation.
+
+In the protocol based implementation, we didn't remove the logic completely from the view controller, although we could have just have easily created view models for each type of 'row' in ProfileViewController. While this leaves view logic in the view controller, we have largely succeeded in making TitleMessageCell reusable.
```

Comment:
I think given that this is going to be a learning tool for very junior developers (among others), we need to be super careful with our examples.  Asking them to see an example and only take away the points we want while ignoring the somewhat bad practices is, I think, asking too much.  So I would like to see an example where the logic *is* removed completely from the VC, so that one takeaway doesn't end up being, "It's fine to leave that sort of logic in the VC."

That said, I also think it is important to get something up and refine it to perfection later.  So I wouldn't necessarily say we have to get it exactly right before it gets published...